protocols/relay/src/v2/relay/handler.rs

```diff
@@ -0,0 +1,892 @@
+                src_peer_id,
+                src_connection_id,
+                inbound_circuit_req: _,
+                dst_handler_notifier: _,
+                dst_stream: _,
+                dst_pending_data: _,
+            } => f
+                .debug_struct("Event::OutboundConnectNegotiated")
+                .field("circuit_id", circuit_id)
+                .field("src_peer_id", src_peer_id)
+                .field("src_connection_id", src_connection_id)
+                .finish(),
+            Event::OutboundConnectNegotiationFailed {
+                circuit_id,
+                src_peer_id,
+                src_connection_id,
+                inbound_circuit_req: _,
+                status,
+            } => f
+                .debug_struct("Event::OutboundConnectNegotiationFailed")
+                .field("circuit_id", circuit_id)
+                .field("src_peer_id", src_peer_id)
+                .field("src_connection_id", src_connection_id)
+                .field("status", status)
+                .finish(),
+            Event::CircuitClosed {
+                circuit_id,
+                dst_peer_id,
+                error,
+            } => f
+                .debug_struct("Event::CircuitClosed")
+                .field("circuit_id", circuit_id)
+                .field("dst_peer_id", dst_peer_id)
+                .field("error", error)
+                .finish(),
+        }
+    }
+}
+
+pub struct Prototype {
+    pub config: Config,
+}
+
+impl IntoProtocolsHandler for Prototype {
+    type Handler = Handler;
+
+    fn into_handler(self, _remote_peer_id: &PeerId, endpoint: &ConnectedPoint) -> Self::Handler {
+        Handler {
+            remote_addr: endpoint.get_remote_address().clone(),
+            config: self.config,
+            queued_events: Default::default(),
+            pending_error: Default::default(),
+            reservation_accept_futures: Default::default(),
+            reservation_deny_futures: Default::default(),
+            circuit_accept_futures: Default::default(),
+            circuit_deny_futures: Default::default(),
+            alive_lend_out_substreams: Default::default(),
+            circuits: Default::default(),
+            active_reservation: Default::default(),
+            keep_alive: KeepAlive::Yes,
+        }
+    }
+
+    fn inbound_protocol(&self) -> <Self::Handler as ProtocolsHandler>::InboundProtocol {
+        inbound_hop::Upgrade {
+            reservation_duration: self.config.reservation_duration,
+            max_circuit_duration: self.config.max_circuit_duration,
+            max_circuit_bytes: self.config.max_circuit_bytes,
+        }
+    }
+}
+
+/// [`ProtocolsHandler`] that manages substreams for a relay on a single
+/// connection with a peer.
+pub struct Handler {
+    remote_addr: Multiaddr,
+
+    /// Static [`Handler`] [`Config`].
+    config: Config,
+
+    /// Queue of events to return when polled.
+    queued_events: VecDeque<
+        ProtocolsHandlerEvent<
```

Comment:
Could we refactor `ProtocolsHandlerEvent` to only take a single generic?
```rust
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum ProtocolsHandlerEvent<THandler: ProtocolsHandler> {
    /// Request a new outbound substream to be opened with the remote.
    OutboundSubstreamRequest {
        /// The protocol(s) to apply on the substream.
        protocol: SubstreamProtocol<THandler::OutboundProtocol, THandler::OutboundOpenInfo>,
    },

    /// Close the connection for the given reason.
    Close(THandler::Errror),

    /// Other event.
    Custom(THandler::OutEvent),
}
```
Would be a bit cleaner.
Or are there any cases where this wouldn't work?

Reply:
Good idea. I suggest exploring this outside of the scope of this pull request. Tracked in https://github.com/libp2p/rust-libp2p/issues/2389.